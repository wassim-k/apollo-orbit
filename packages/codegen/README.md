# Apollo Orbit (Codegen)

Documentation coming soon...
