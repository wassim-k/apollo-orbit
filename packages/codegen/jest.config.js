import { dirname } from 'path';
import { fileURLToPath } from 'url';
import configureProject from '../../jest.project.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export default {
  ...configureProject(__dirname),
  resolver: '../../node_modules/bob-the-bundler/jest-resolver.cjs',
};
