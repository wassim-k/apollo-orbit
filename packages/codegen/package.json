{
  "name": "@apollo-orbit/codegen",
  "version": "2.0.1",
  "description": "GraphQL Code Generator plugin for Apollo Orbit",
  "license": "MIT",
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "git+https://github.com/wassim-k/apollo-orbit.git"
  },
  "scripts": {
    "build": "bob build",
    "clean": "rimraf dist",
    "prebuild": "yarn clean"
  },
  "peerDependencies": {
    "graphql": ">=0.11"
  },
  "dependencies": {
    "@graphql-codegen/plugin-helpers": "^5.1.1",
    "@graphql-codegen/visitor-plugin-common": "^5.8.0",
    "auto-bind": "~4.0.0",
    "tslib": "^2.8.1"
  },
  "devDependencies": {
    "@graphql-codegen/testing": "^3.0.4",
    "bob-the-bundler": "^7.0.1",
    "jest-diff": "^29.0.0"
  },
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "exports": {
    ".": {
      "require": {
        "types": "./dist/typings/index.d.cts",
        "default": "./dist/cjs/index.js"
      },
      "import": {
        "types": "./dist/typings/index.d.ts",
        "default": "./dist/esm/index.js"
      },
      "default": {
        "types": "./dist/typings/index.d.ts",
        "default": "./dist/esm/index.js"
      }
    },
    "./core": {
      "require": {
        "types": "./dist/typings/core/index.d.cts",
        "default": "./dist/cjs/core/index.js"
      },
      "import": {
        "types": "./dist/typings/core/index.d.ts",
        "default": "./dist/esm/core/index.js"
      },
      "default": {
        "types": "./dist/typings/core/index.d.ts",
        "default": "./dist/esm/core/index.js"
      }
    },
    "./angular": {
      "require": {
        "types": "./dist/typings/angular/index.d.cts",
        "default": "./dist/cjs/angular/index.js"
      },
      "import": {
        "types": "./dist/typings/angular/index.d.ts",
        "default": "./dist/esm/angular/index.js"
      },
      "default": {
        "types": "./dist/typings/angular/index.d.ts",
        "default": "./dist/esm/angular/index.js"
      }
    },
    "./package.json": "./package.json"
  },
  "typings": "dist/typings/index.d.ts",
  "typescript": {
    "definition": "dist/typings/index.d.ts"
  },
  "publishConfig": {
    "directory": "dist",
    "access": "public"
  },
  "type": "module",
  "engines": {
    "node": ">=16"
  }
}
