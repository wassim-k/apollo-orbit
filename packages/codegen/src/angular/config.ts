/* eslint-disable @typescript-eslint/no-empty-object-type */

import { ApolloOrbitPluginConfig, ApolloOrbitRawPluginConfig } from '../core/config.js';

export interface ApolloOrbitAngularConfig extends ApolloOrbitPluginConfig {
}

export interface ApolloOrbitAngularRawConfig extends ApolloOrbitRawPluginConfig {
}
