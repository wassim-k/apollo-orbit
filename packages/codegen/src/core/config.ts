import { ClientSideBasePluginConfig, RawClientSideBasePluginConfig } from '@graphql-codegen/visitor-plugin-common';

export interface ApolloOrbitPluginConfig extends ClientSideBasePluginConfig {
  querySuffix: string;
  mutationSuffix: string;
  subscriptionSuffix: string;
  importPath: string;
  gqlFunction: boolean;
  gqlVariablesAsFunction: boolean;
  documentFieldName?: {
    query?: string;
    mutation?: string;
    subscription?: string;
  };
}

export interface ApolloOrbitRawPluginConfig extends RawClientSideBasePluginConfig {
  /**
   * @name querySuffix
   * @type string
   * @description Allows to define a custom suffix for query operations.
   * @default 'Query'
   *
   * @example graphql.macro
   * ```yml
   * config:
   *   querySuffix: 'Query'
   * ```
   */
  querySuffix?: string;

  /**
   * @name mutationSuffix
   * @type string
   * @description Allows to define a custom suffix for mutation operations.
   * @default 'Mutation'
   *
   * @example graphql.macro
   * ```yml
   * config:
   *   mutationSuffix: 'Mutation'
   * ```
   */
  mutationSuffix?: string;

  /**
   * @name subscriptionSuffix
   * @type string
   * @description Allows to define a custom suffix for Subscription operations.
   * @default 'Subscription'
   *
   * @example graphql.macro
   * ```yml
   * config:
   *   subscriptionSuffix: 'Subscription'
   * ```
   */
  subscriptionSuffix?: string;

  /**
   * @name gqlFunction
   * @type boolean
   * @description Generate a gql function for each operation.
   * @default true
   *
   * @example graphql.macro
   * ```yml
   * config:
   *  gqlFunction: true
   * ```
   */
  gqlFunction?: boolean;

  /**
   * @name gqlVariablesAsFunction
   * @type boolean
   * @description Generate an gql function overload with variables as a function parameter.
   * @default false
   *
   * @example graphql.macro
   * ```yml
   * config:
   *  gqlVariablesAsFunction: true
   * ```
   */
  gqlVariablesAsFunction?: boolean;

  /**
   * @name documentFieldName
   * @type object
   * @description Allows to define a custom field name for the document object in the generated code.
   * @default false
   *
   * @example graphql.macro
   * ```yml
   * config:
   *  documentFieldName:
   *    query: 'query',
   *    mutation: 'mutation',
   *    subscription: 'subscription',
   * ```
   */
  documentFieldName?: {
    query?: string;
    mutation?: string;
    subscription?: string;
  };
}
