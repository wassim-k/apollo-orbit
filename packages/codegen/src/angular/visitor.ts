import { Types } from '@graphql-codegen/plugin-helpers';
import { getConfigValue, LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import { GraphQLSchema, OperationDefinitionNode } from 'graphql';
import { ApolloOrbitVisitor, OperationType } from '../core/visitor.js';
import { ApolloOrbitAngularConfig, ApolloOrbitAngularRawConfig } from './config.js';

export class ApolloOrbitAngularVisitor extends ApolloOrbitVisitor<ApolloOrbitAngularRawConfig, ApolloOrbitAngularConfig> {
  public constructor(
    schema: GraphQLSchema,
    fragments: Array<LoadedFragment>,
    rawConfig: ApolloOrbitAngularRawConfig,
    documents?: Array<Types.DocumentFile>
  ) {
    super(
      schema,
      fragments,
      rawConfig,
      {
        importPath: '@apollo-orbit/angular',
        gqlFunction: getConfigValue(rawConfig.gqlFunction, true),
        gqlVariablesAsFunction: getConfigValue(rawConfig.gqlVariablesAsFunction, true),
        documentFieldName: getConfigValue(rawConfig.documentFieldName, {
          query: 'query',
          mutation: 'mutation',
          subscription: 'subscription'
        })
      } as ApolloOrbitAngularConfig,
      documents
    );
  }

  protected buildOperation(
    node: OperationDefinitionNode,
    documentVariableName: string,
    operationType: OperationType,
    operationResultType: string,
    operationVariablesTypes: string,
    hasRequiredVariables: boolean
  ): string {
    return super.buildOperation(
      node,
      documentVariableName,
      operationType,
      operationResultType,
      operationVariablesTypes,
      hasRequiredVariables
    );
  }
}
