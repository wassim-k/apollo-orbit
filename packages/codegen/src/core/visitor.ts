import { Types } from '@graphql-codegen/plugin-helpers';
import { ClientSideBaseVisitor, DocumentMode, getConfigValue, LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import autoBind from 'auto-bind';
import { constantCase, pascalCase } from 'change-case-all';
import { FragmentDefinitionNode, GraphQLSchema, OperationDefinitionNode } from 'graphql';
import { Importer } from '../internal/importer.js';
import { ApolloOrbitPluginConfig, ApolloOrbitRawPluginConfig } from './config.js';

export type OperationType = 'Query' | 'Mutation' | 'Subscription';

const DOCUMENT_VARIABLE_OPERATION_SUFFIX = '[OPERATION]';

export class ApolloOrbitVisitor<
  TRawConfig extends ApolloOrbitRawPluginConfig = ApolloOrbitRawPluginConfig,
  TPluginConfig extends ApolloOrbitPluginConfig = ApolloOrbitPluginConfig
> extends ClientSideBaseVisitor<TRawConfig, TPluginConfig> {
  private readonly importer = new Importer();

  public constructor(
    schema: GraphQLSchema,
    fragments: Array<LoadedFragment>,
    rawConfig: TRawConfig,
    additionalConfig: Partial<TPluginConfig>,
    documents?: Array<Types.DocumentFile>
  ) {
    super(
      schema,
      fragments,
      {
        gqlImport: `${additionalConfig.importPath ?? '@apollo/client'}#gql`,
        documentNodeImport: `${additionalConfig.importPath ?? '@apollo/client'}#TypedDocumentNode`,
        dedupeOperationSuffix: true,
        documentVariableSuffix: getConfigValue(rawConfig.documentVariableSuffix, DOCUMENT_VARIABLE_OPERATION_SUFFIX),
        ...rawConfig
      },
      {
        querySuffix: getConfigValue(rawConfig.querySuffix, 'Query'),
        mutationSuffix: getConfigValue(rawConfig.mutationSuffix, 'Mutation'),
        subscriptionSuffix: getConfigValue(rawConfig.subscriptionSuffix, 'Subscription'),
        gqlFunction: getConfigValue(rawConfig.gqlFunction, false),
        gqlVariablesAsFunction: getConfigValue(rawConfig.gqlVariablesAsFunction, false),
        ...additionalConfig
      },
      documents);

    autoBind(this);

    if (this.config.documentMode === undefined || this.config.documentMode === DocumentMode.graphQLTag) {
      const documentNodeImport = this._parseImport(this.config.documentNodeImport || 'graphql#DocumentNode');
      this.registerImport(documentNodeImport.propName, documentNodeImport.moduleName as string, 'DocumentNode');
    }
  }

  public getImports(): Array<string> {
    return [
      ...super.getImports(),
      ...this.importer.getImports()
    ];
  }

  public getOperationVariableName(node: OperationDefinitionNode): string {
    return constantCase(this.convertName(node, {
      suffix: this.config.documentVariableSuffix === DOCUMENT_VARIABLE_OPERATION_SUFFIX
        ? this.getOperationSuffix(node, pascalCase(node.operation))
        : this.config.documentVariableSuffix,
      prefix: this.config.documentVariablePrefix
    }));
  }

  protected buildOperation(
    node: OperationDefinitionNode,
    documentVariableName: string,
    operationType: OperationType,
    _dataType: string,
    variablesTypes: string,
    hasRequiredVariables: boolean
  ): string {
    if (!this.config.gqlFunction) return '';

    const functionName = `gql${this.convertName(node, { suffix: this.operationSuffix(operationType) })}`;
    const documentNodeVariable = this.getDocumentNodeVariable(node, documentVariableName);
    const documentFieldName = this.getDocumentFieldName(operationType);

    const content: Array<string> = [];

    switch (operationType) {
      case 'Query':
      case 'Subscription':
        if (this.hasVariables(node)) {
          const isOptional = !hasRequiredVariables;

          content.push('');

          if (!this.config.gqlVariablesAsFunction) {
            content.push(`export function ${functionName}(variables${isOptional ? '?' : ''}: ${variablesTypes}) {
  return {
    ${documentFieldName}: ${documentNodeVariable},
    variables
  };
}`);
          } else {
            if (isOptional) {
              content.push(`export function ${functionName}(): ${this.createDocumentOnlyReturnType(documentFieldName, documentNodeVariable)};`);
            }
            content.push(`export function ${functionName}(variables${isOptional ? '?' : ''}: ${variablesTypes}): ${this.createDocumentReturnType(documentFieldName, documentNodeVariable)};
export function ${functionName}(variables: () => ${variablesTypes}${isOptional ? ' | undefined' : ''}): ${this.createDocumentReturnType(documentFieldName, documentNodeVariable)};
export function ${functionName}(variables${isOptional ? '?' : ''}: any): any {
  return {
    ${documentFieldName}: ${documentNodeVariable},
    variables
  };
}`);
          }
        } else {
          content.push(`
export function ${functionName}(): ${this.createDocumentOnlyReturnType(documentFieldName, documentNodeVariable)} {
  return {
    ${documentFieldName}: ${documentNodeVariable}
  };
}`);
        }
        break;

      case 'Mutation':
        if (this.hasVariables(node)) {
          const isOptional = !hasRequiredVariables;

          content.push('');

          content.push(`export function ${functionName}(variables${isOptional ? '?' : ''}: ${variablesTypes}): ${this.createDocumentReturnType(documentFieldName, documentNodeVariable)} {
  return {
    ${documentFieldName}: ${documentNodeVariable},
    variables
  };
}`);
        } else {
          content.push(`
export function ${functionName}(): ${this.createDocumentOnlyReturnType(documentFieldName, documentNodeVariable)} {
  return {
    ${documentFieldName}: ${documentNodeVariable}
  };
}`);
        }
        break;
      default: throw new Error(`Unsupported operation type: ${operationType}`);
    }

    content.push('');
    return content.join('\n');
  }

  protected getDocumentNodeSignature(resultType: string, variablesTypes: string, node: FragmentDefinitionNode | OperationDefinitionNode): string {
    if (
      this.config.documentMode === DocumentMode.documentNode ||
      this.config.documentMode === DocumentMode.documentNodeImportFragments ||
      this.config.documentMode === DocumentMode.graphQLTag
    ) {
      return ` as DocumentNode<${resultType}, ${variablesTypes}>`;
    }

    return super.getDocumentNodeSignature(resultType, variablesTypes, node);
  }

  protected registerImport(identifier: string, from?: string, as?: string): void {
    this.importer.registerImport(identifier, from ?? this.config.importPath, as);
  }

  private hasVariables(node: OperationDefinitionNode): boolean {
    return node.variableDefinitions !== undefined && node.variableDefinitions.length > 0;
  }

  private getDocumentNodeVariable(node: OperationDefinitionNode, documentVariableName: string): string {
    if (!node.name) throw new Error('Operation node must have a name.');
    return this.config.documentMode === DocumentMode.external ? `Operations.${node.name.value}` : documentVariableName;
  }

  private operationSuffix(operationType: OperationType): string {
    const { querySuffix, mutationSuffix, subscriptionSuffix } = this.config;
    const suffixByOperation = {
      Query: querySuffix,
      Mutation: mutationSuffix,
      Subscription: subscriptionSuffix
    };
    return suffixByOperation[operationType];
  }

  private getDocumentFieldName(operationType: OperationType): string {
    switch (operationType) {
      case 'Query': return this.config.documentFieldName?.query ?? 'query';
      case 'Mutation': return this.config.documentFieldName?.mutation ?? 'mutation';
      case 'Subscription': return this.config.documentFieldName?.subscription ?? 'subscription';
      default: throw new Error(`Unsupported operation type: ${operationType}`);
    }
  }

  private createDocumentOnlyReturnType(documentFieldName: string, documentNodeVariable: string): string {
    return `{ ${documentFieldName}: typeof ${documentNodeVariable} }`;
  }

  private createDocumentReturnType(documentFieldName: string, documentNodeVariable: string): string {
    return `{ ${documentFieldName}: typeof ${documentNodeVariable}, variables: typeof variables }`;
  }
}
