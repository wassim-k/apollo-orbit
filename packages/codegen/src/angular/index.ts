import { PluginFunction, PluginValidateFn, Types } from '@graphql-codegen/plugin-helpers';
import { DocumentMode } from '@graphql-codegen/visitor-plugin-common';
import { GraphQLSchema, visit } from 'graphql';
import { extname } from 'path';
import { prepareDocuments } from '../internal/prepareDocuments.js';
import { ApolloOrbitAngularConfig, ApolloOrbitAngularRawConfig } from './config.js';
import { ApolloOrbitAngularVisitor } from './visitor.js';

export const plugin: PluginFunction<ApolloOrbitAngularRawConfig> = (
  schema: GraphQLSchema,
  documents: Array<Types.DocumentFile>,
  config
) => {
  const [allFragments, allAst] = prepareDocuments(documents, config.externalFragments);
  const visitor = new ApolloOrbitAngularVisitor(schema, allFragments, config, documents);
  const visitorResult = visit(allAst, visitor);
  return {
    prepend: visitor.getImports(),
    content: [visitor.fragments, ...visitorResult.definitions.filter((t: unknown): t is string => typeof t === 'string')]
      .filter(definition => typeof definition !== 'undefined')
      .join('\n')
  };
};

export const validate: PluginValidateFn = (
  _schema: GraphQLSchema,
  _documents: Array<Types.DocumentFile>,
  config: ApolloOrbitAngularConfig,
  outputFile: string
): void => {
  if (config.documentMode === DocumentMode.string) {
    throw new Error('Plugin "@apollo-orbit/codegen" does not support \'documentMode: string\' configuration');
  }

  if (extname(outputFile) !== '.ts') {
    throw new Error('Plugin "@apollo-orbit/codegen" requires extension to be ".ts"');
  }
};
