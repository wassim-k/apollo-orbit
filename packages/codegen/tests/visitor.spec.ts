import { plugin as angularPlugin } from '@apollo-orbit/codegen/angular';
import { plugin as corePlugin } from '@apollo-orbit/codegen/core';
import { gql } from '@apollo/client';
import { mergeOutputs, Types } from '@graphql-codegen/plugin-helpers';
import { validateTs } from '@graphql-codegen/testing';
import { plugin as tsPlugin } from '@graphql-codegen/typescript';
import { plugin as tsDocumentsPlugin } from '@graphql-codegen/typescript-operations';
import fs from 'fs';
import { buildSchema, GraphQLSchema } from 'graphql';
import path from 'path';

describe('Codegen', () => {
  const schemaPath = path.join(__dirname, './graphql/schema.graphql');
  const schemaDocument: string = fs.readFileSync(schemaPath, 'utf8');
  const schema = buildSchema(schemaDocument);

  const validateTypeScript = async (
    output: Types.PluginOutput,
    testSchema: GraphQLSchema,
    documents: Array<Types.DocumentFile>,
    config: any
  ): Promise<void> => {
    const tsOutput = await tsPlugin(testSchema, documents, config, { outputFile: '' });
    const tsDocumentsOutput = await tsDocumentsPlugin(testSchema, documents, config, { outputFile: '' });
    const merged = mergeOutputs([tsOutput, tsDocumentsOutput, output]);
    validateTs(merged, undefined, true);
  };

  describe('Core', () => {
    it('should codegen', async () => {
      const books = gql`
        query Books {
          books {
            id
          }
        }`;

      const book = gql`
        query Book($id: ID!) {
          book(id: $id) {
            id
          }
        }`;

      const addBook = gql`
        mutation AddBook($book: BookInput!) {
          addBook(book: $book) {
            id
          }
        }`;

      const newBook = gql`
        subscription NewBook {
          newBook {
            id
          }
        }`;

      const docs = [
        { location: '', document: books },
        { location: '', document: book },
        { location: '', document: addBook },
        { location: '', document: newBook }
      ];
      const content = (await corePlugin(
        schema,
        docs,
        {},
        {
          outputFile: 'graphql.ts'
        }
      )) as Types.ComplexPluginOutput;

      await validateTypeScript(content, schema, docs, {});
      expect(content).toMatchSnapshot();
    });
  });

  describe('Angular', () => {
    it('should codegen', async () => {
      const books = gql`
      query Books {
        books {
          id
        }
      }`;

      const book = gql`
      query Book($id: ID!) {
        book(id: $id) {
          id
        }
      }`;

      const addBook = gql`
      mutation AddBook($book: BookInput!) {
        addBook(book: $book) {
          id
        }
      }`;

      const newBook = gql`
      subscription NewBook {
        newBook {
          id
        }
      }`;

      const docs = [
        { location: '', document: books },
        { location: '', document: book },
        { location: '', document: addBook },
        { location: '', document: newBook }
      ];
      const content = (await angularPlugin(
        schema,
        docs,
        {},
        {
          outputFile: 'graphql.ts'
        }
      )) as Types.ComplexPluginOutput;

      await validateTypeScript(content, schema, docs, {});
      expect(content).toMatchSnapshot();
    });
  });
});
