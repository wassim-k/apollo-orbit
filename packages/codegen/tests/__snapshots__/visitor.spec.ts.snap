// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Codegen Angular should codegen 1`] = `
{
  "content": "
export const BOOKS_QUERY = gql\`
    query Books {
  books {
    id
  }
}
    \` as DocumentNode<BooksQuery, BooksQueryVariables>;

export function gqlBooksQuery(): { query: typeof BOOKS_QUERY } {
  return {
    query: BOOKS_QUERY
  };
}

export const BOOK_QUERY = gql\`
    query Book($id: ID!) {
  book(id: $id) {
    id
  }
}
    \` as DocumentNode<BookQuery, BookQueryVariables>;

export function gqlBookQuery(variables: BookQueryVariables): { query: typeof BOOK_QUERY, variables: typeof variables };
export function gqlBookQuery(variables: () => BookQueryVariables): { query: typeof BOOK_QUERY, variables: typeof variables };
export function gqlBookQuery(variables: any): any {
  return {
    query: BOOK_QUERY,
    variables
  };
}

export const ADD_BOOK_MUTATION = gql\`
    mutation AddBook($book: BookInput!) {
  addBook(book: $book) {
    id
  }
}
    \` as DocumentNode<AddBookMutation, AddBookMutationVariables>;

export function gqlAddBookMutation(variables: AddBookMutationVariables): { mutation: typeof ADD_BOOK_MUTATION, variables: typeof variables } {
  return {
    mutation: ADD_BOOK_MUTATION,
    variables
  };
}

export const NEW_BOOK_SUBSCRIPTION = gql\`
    subscription NewBook {
  newBook {
    id
  }
}
    \` as DocumentNode<NewBookSubscription, NewBookSubscriptionVariables>;

export function gqlNewBookSubscription(): { subscription: typeof NEW_BOOK_SUBSCRIPTION } {
  return {
    subscription: NEW_BOOK_SUBSCRIPTION
  };
}
",
  "prepend": [
    "import { gql } from '@apollo-orbit/angular';",
    "import { TypedDocumentNode as DocumentNode } from '@apollo-orbit/angular';",
  ],
}
`;

exports[`Codegen Core should codegen 1`] = `
{
  "content": "
export const BOOKS_QUERY = gql\`
    query Books {
  books {
    id
  }
}
    \` as DocumentNode<BooksQuery, BooksQueryVariables>;
export const BOOK_QUERY = gql\`
    query Book($id: ID!) {
  book(id: $id) {
    id
  }
}
    \` as DocumentNode<BookQuery, BookQueryVariables>;
export const ADD_BOOK_MUTATION = gql\`
    mutation AddBook($book: BookInput!) {
  addBook(book: $book) {
    id
  }
}
    \` as DocumentNode<AddBookMutation, AddBookMutationVariables>;
export const NEW_BOOK_SUBSCRIPTION = gql\`
    subscription NewBook {
  newBook {
    id
  }
}
    \` as DocumentNode<NewBookSubscription, NewBookSubscriptionVariables>;",
  "prepend": [
    "import { gql } from '@apollo/client';",
    "import { TypedDocumentNode as DocumentNode } from '@apollo/client';",
  ],
}
`;
